### ISSUE TITLE
Multiple reports

### ISSUE THREAD

**Issue by lpefferkorn** [lpefferkorn/cfe-profiler#5]
Steps to give me more information (you will need gdb):

```
$ gdb --args  ~/tmp/cfe341/bin/cf-agent -KI
Reading symbols from /home/loic/tmp/cfe341/bin/cf-agent...done.
(gdb) break GenericDeInitialize
Breakpoint 1 at 0x406bf0
(gdb) command 1
Type commands for breakpoint(s) 1, one per line.
End with a line saying just "end".
>backtrace
>continue
>end
(gdb) run
```

Then, for each GenericDeInitialize() call, there should be the backtrace, which is going to help me to find out why GenericDeInitialize() is called

**Comment by neilhwatson** [comment]
Do I have to set LD_PRELOAD somewhere in this?  If so, where?

**Comment by lpefferkorn** [comment]
You can run the gdb session without LD_PRELOAD

**Comment by neilhwatson** [comment]
[root@ltipc682 ~]# gdb --args /var/cfengine/bin/cf-agent -K
GNU gdb (GDB) Red Hat Enterprise Linux (7.2-56.el6)
Copyright (C) 2010 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later http://gnu.org/licenses/gpl.html
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
http://www.gnu.org/software/gdb/bugs/...
Reading symbols from /var/cfengine/bin/cf-agent...(no debugging symbols found)...done.
(gdb) back
No stack.

**Comment by lpefferkorn** [comment]
I wrote a fix on atexit branch https://github.com/lpefferkorn/cfe-profiler/tree/atexit

Does it fix this behavior ?

**Comment by neilhwatson** [comment]
Sorry, no change.

**Comment by lpefferkorn** [comment]
I saw that you typed "back" in gdb, but in fact this is "breakpoint" command. I have updated my previous message to include full gdb commands

**Comment by neilhwatson** [comment]
Sorry Loic, I don't understand what you need. Assume I know next to nothing about gdb.

**Comment by neilhwatson** [comment]
I should specify my troubles.  This agent runs is verbose mode and it's a big policy. I cannot capture on the screen all the information that flies by. Is there a way to log gdb's output?

**Comment by lpefferkorn** [comment]
Yes, at the gdb prompt, type "set logging on", you should have the following output:

```
(gdb) set logging on
Copying output to gdb.txt.
(gdb) show logging
Currently logging to "gdb.txt".
Logs will be appended to the log file.
Output is being logged and displayed.
```

Future output will be both displayed and recorded to gdb.txt file

And to avoid "type enter to continue", the command is "set pagination off"

```
(gdb) set pagination off
```

All together:

```
$ gdb --args  ~/tmp/cfe341/bin/cf-agent -KI
Reading symbols from /home/loic/tmp/cfe341/bin/cf-agent...done.
(gdb) set logging on
Copying output to gdb.txt.
(gdb) set pagination off
(gdb) break GenericDeInitialize
Breakpoint 1 at 0x406bf0
(gdb) command 1
Type commands for breakpoint(s) 1, one per line.
End with a line saying just "end".
>backtrace
>continue
>end
(gdb) run
```

(gdb) means command prompt, where you have to type the commands I have listed

**Comment by lpefferkorn** [comment]
Easier using gdb batch mode :)

Create a file /tmp/gdb.script with the following contents

```
set verbose off
set logging redirect on
set logging on
set pagination off
break GenericDeInitialize
command 1
  backtrace
  continue
end
run
```

And run

```
 gdb --batch --command=/tmp/gdb.script --args ~/tmp/cfe341/bin/cf-agent -KI
```

And everything I need is going to be written to gdb.txt

**Comment by lpefferkorn** [comment]
If you need help with gdb, we can talk on irc, lpn on Freenode

**Comment by neilhwatson** [comment]
Breakpoint 1 at 0x407dc8
[Thread debugging using libthread_db enabled]
Detaching after fork from child process 29447.
Detaching after fork from child process 29448.
Detaching after fork from child process 29451.
Detaching after fork from child process 29452.
Detaching after fork from child process 29453.
Detaching after fork from child process 29456.
Detaching after fork from child process 29458.
Detaching after fork from child process 29459.
Detaching after fork from child process 29460.
Detaching after fork from child process 29461.
Detaching after fork from child process 29462.
Detaching after fork from child process 29463.
Detaching after fork from child process 29464.
Detaching after fork from child process 29465.
Detaching after fork from child process 29466.
Detaching after fork from child process 29467.
Detaching after fork from child process 29468.
Detaching after fork from child process 29469.
Detaching after fork from child process 29470.
Detaching after fork from child process 29471.
Detaching after fork from child process 29472.
Detaching after fork from child process 29473.
Detaching after fork from child process 29474.
Detaching after fork from child process 29475.
Detaching after fork from child process 29476.
Detaching after fork from child process 29477.
Detaching after fork from child process 29478.
Detaching after fork from child process 29479.
Detaching after fork from child process 29480.
Detaching after fork from child process 29481.
Detaching after fork from child process 29482.
Detaching after fork from child process 29483.
Detaching after fork from child process 29484.
Detaching after fork from child process 29485.
Detaching after fork from child process 29486.
Detaching after fork from child process 29487.
Detaching after fork from child process 29488.
Detaching after fork from child process 29489.
Detaching after fork from child process 29490.
Detaching after fork from child process 29491.
Detaching after fork from child process 29492.
Detaching after fork from child process 29493.
Detaching after fork from child process 29494.
Detaching after fork from child process 29495.
Detaching after fork from child process 29496.
Detaching after fork from child process 29497.
Detaching after fork from child process 29498.
Detaching after fork from child process 29501.
Detaching after fork from child process 29502.
Detaching after fork from child process 29503.
Detaching after fork from child process 29504.
Detaching after fork from child process 29505.
Detaching after fork from child process 29506.
Detaching after fork from child process 29507.
Detaching after fork from child process 29508.
Detaching after fork from child process 29509.
Detaching after fork from child process 29510.
Detaching after fork from child process 29511.
Detaching after fork from child process 29512.
Detaching after fork from child process 29513.
Detaching after fork from child process 29514.
Detaching after fork from child process 29515.
Detaching after fork from child process 29516.
Detaching after fork from child process 29517.
Detaching after fork from child process 29518.
Detaching after fork from child process 29519.
Detaching after fork from child process 29520.
Detaching after fork from child process 29521.
Detaching after fork from child process 29522.
Detaching after fork from child process 29523.
Detaching after fork from child process 29524.
Detaching after fork from child process 29525.
Detaching after fork from child process 29526.
Detaching after fork from child process 29555.
Detaching after fork from child process 29556.
Detaching after fork from child process 29558.
Detaching after fork from child process 29561.
Detaching after fork from child process 29590.
Detaching after fork from child process 29591.
Detaching after fork from child process 29593.
Detaching after fork from child process 29596.
Detaching after fork from child process 29625.
Detaching after fork from child process 29626.
Detaching after fork from child process 29628.
Detaching after fork from child process 29631.
Detaching after fork from child process 29632.
Detaching after fork from child process 29633.
Detaching after fork from child process 29634.
Detaching after fork from child process 29635.
Detaching after fork from child process 29636.
Detaching after fork from child process 29637.
Detaching after fork from child process 29638.
Detaching after fork from child process 29639.
Detaching after fork from child process 29640.
Detaching after fork from child process 29641.
Detaching after fork from child process 29642.
Detaching after fork from child process 29643.
Detaching after fork from child process 29644.
Detaching after fork from child process 29645.
Detaching after fork from child process 29646.
Detaching after fork from child process 29647.
Detaching after fork from child process 29648.
Detaching after fork from child process 29649.
Detaching after fork from child process 29650.
Detaching after fork from child process 29651.
Detaching after fork from child process 29652.
Detaching after fork from child process 29653.
Detaching after fork from child process 29654.
Detaching after fork from child process 29655.
Detaching after fork from child process 29656.
Detaching after fork from child process 29657.
Detaching after fork from child process 29658
Detaching after fork from child process 29659.
Detaching after fork from child process 29660.
Detaching after fork from child process 29661.
Detaching after fork from child process 29662.
Detaching after fork from child process 29663.
Detaching after fork from child process 29664.
Detaching after fork from child process 29665.
Detaching after fork from child process 29666.
Detaching after fork from child process 29667.
Detaching after fork from child process 29668.
Detaching after fork from child process 29669.
Detaching after fork from child process 29670.
Detaching after fork from child process 29671.
Detaching after fork from child process 29672.
Detaching after fork from child process 29673.
Detaching after fork from child process 29674.
Detaching after fork from child process 29675.
Detaching after fork from child process 29676.
Detaching after fork from child process 29677.
Detaching after fork from child process 29678.
Detaching after fork from child process 29679.
Detaching after fork from child process 29680.
Detaching after fork from child process 29681.
Detaching after fork from child process 29682.
Detaching after fork from child process 29683.
Detaching after fork from child process 29684.
Detaching after fork from child process 29685.
Detaching after fork from child process 29686.
Detaching after fork from child process 29687.
Detaching after fork from child process 29688.
Detaching after fork from child process 29689.
Detaching after fork from child process 29690.
Detaching after fork from child process 29691.
Detaching after fork from child process 29692.
Detaching after fork from child process 29693.
Detaching after fork from child process 29694.
Detaching after fork from child process 29695.
Detaching after fork from child process 29696.
Detaching after fork from child process 29697.
Detaching after fork from child process 29698.
Detaching after fork from child process 29699.
Detaching after fork from child process 29700.
Detaching after fork from child process 29701.
Detaching after fork from child process 29702.
Detaching after fork from child process 29703.
Detaching after fork from child process 29704.
Detaching after fork from child process 29705.
Detaching after fork from child process 29706.
Detaching after fork from child process 29707.
Detaching after fork from child process 29708.
Detaching after fork from child process 29709.
Detaching after fork from child process 29710.
Detaching after fork from child process 29711.
Detaching after fork from child process 29712.
Detaching after fork from child process 29713.
Detaching after fork from child process 29714.
Detaching after fork from child process 29715.
Detaching after fork from child process 29716.
Detaching after fork from child process 29717.
Detaching after fork from child process 29718.
Detaching after fork from child process 29719.
Detaching after fork from child process 29720.
Detaching after fork from child process 29721.
Detaching after fork from child process 29722.
Detaching after fork from child process 29723.
Detaching after fork from child process 29724.
Detaching after fork from child process 29725.
Detaching after fork from child process 29726.
Detaching after fork from child process 29727.
Detaching after fork from child process 29728.
Detaching after fork from child process 29729.
Detaching after fork from child process 29730.
Detaching after fork from child process 29731.
Detaching after fork from child process 29732.
Detaching after fork from child process 29733.
Detaching after fork from child process 29734.
Detaching after fork from child process 29735.
Detaching after fork from child process 29736.
Detaching after fork from child process 29737.
Detaching after fork from child process 29738.
Detaching after fork from child process 29739.
Detaching after fork from child process 29740.
Detaching after fork from child process 29741.
Detaching after fork from child process 29742.
Detaching after fork from child process 29743.
Detaching after fork from child process 29744.
Detaching after fork from child process 29745.
Detaching after fork from child process 29746.
Detaching after fork from child process 29747.
Detaching after fork from child process 29748.
Detaching after fork from child process 29749.
Detaching after fork from child process 29750.
Detaching after fork from child process 29751.
Detaching after fork from child process 29752.
Detaching after fork from child process 29753.
Detaching after fork from child process 29754.
Detaching after fork from child process 29755.
Detaching after fork from child process 29756.
Detaching after fork from child process 29757.
Detaching after fork from child process 29758.
Detaching after fork from child process 29759.
Detaching after fork from child process 29761.
Detaching after fork from child process 29763.
Detaching after fork from child process 29766.
Detaching after fork from child process 29769.
Detaching after fork from child process 29772.
Detaching after fork from child process 29775.
Detaching after fork from child process 29776.
Detaching after fork from child process 29777.
Detaching after fork from child process 29778.
Detaching after fork from child process 29781.
Detaching after fork from child process 29782.
Detaching after fork from child process 29785.
Detaching after fork from child process 29788.
Detaching after fork from child process 29791.
Detaching after fork from child process 29794.
Detaching after fork from child process 29795.
Detaching after fork from child process 29797.
Detaching after fork from child process 29799.
Detaching after fork from child process 29800.
Detaching after fork from child process 29802.
Detaching after fork from child process 29804.
Detaching after fork from child process 29806.
Detaching after fork from child process 29807.
Detaching after fork from child process 29809.
Detaching after fork from child process 29811.
Detaching after fork from child process 29813.
Detaching after fork from child process 29814.
Detaching after fork from child process 29816.
Detaching after fork from child process 29818.
Detaching after fork from child process 29819.
Detaching after fork from child process 29820.
Detaching after fork from child process 29821.
Detaching after fork from child process 29822.
Detaching after fork from child process 29824.
Detaching after fork from child process 29826.

Breakpoint 1, GenericDeInitialize () at generic_agent.c:262
262     generic_agent.c: No such file or directory.
        in generic_agent.c
#0  GenericDeInitialize () at generic_agent.c:262
#1  0x000000000040a51e in main ()

Program exited normally.

**Comment by lpefferkorn** [comment]
@neilhwatson Would you mind starting again the gdb processus with this updated gdb.script ?

```
set verbose off
set logging redirect on
set logging on
set pagination off
break GenericDeInitialize
command 1
  backtrace
  info threads
  continue
end
run
```

**Comment by lpefferkorn** [comment]
@neilhwatson Does my latest commit in https://github.com/lpefferkorn/cfe-profiler/tree/atexit branch fix the issue ?

**Comment by lpefferkorn** [comment]
Please check https://github.com/lpefferkorn/cfe-profiler/tree/dev instead :)

**Comment by lpefferkorn** [comment]
Fixed in version 0.2 https://github.com/lpefferkorn/cfe-profiler/releases/tag/v0.2
